doc/adr/0016-state-machines-for-UI-state.md

```diff
@@ -0,0 +1,59 @@
+# 0016 State machines for UI state management
+
+Date: Unknown
+
+## Status
+
+Accepted
+
+## Context
+
+Modern (JavaScript-based) User Interfaces have many components and lots of
+concurrent actions going on.  Due to the combination and number of possible
+actions being triggered, the integration between components has become
+increasingly complex over the years -- especially when considering
+asynchronicity and error handling.
+
+The JavaScript/front-end development community has increasingly been using
+[state machines](https://en.wikipedia.org/wiki/Finite-state_machine) and
+[state charts](https://en.wikipedia.org/wiki/UML_state_machine) to solve this
+issue of complexity.
+
+[XState](https://stately.ai/docs/xstate) is the - by far - largest state
+machine engine in the JavaScript ecosystem, featuring over 1 million downloads
+per week (ref: 2023-12-02).  It implements the *state chart* state machine
+type.  XState however is an extensive library with an extensive 'DSL' to
+describe state machines.  There are many alternatives implementing state charts
+or state machines that are less extensive.  One such example is [robot3](https://thisrobot.life/)
+which is extremely small, supports sub-state machines and features integration
+for React, PReact and Lit.  Its DSL is similarly small and thereby easy to
+read.  None of the integrations are of interest for LedgerSMB, but they
+indicate interest from a wider community.
+
+Considering that with its move to Vue, LedgerSMB is going in the same
+direction as other projects -- featuring a heavy (JavaScript-based) client
+application -- the project similarly needs a way to control the expected
+complexity.
+
+## Decision
+
+The web-based client application will use state machines to make state
+management as well as state transition explicit.  Similar to UI composition
+with components, will different parts of the UI be governed by their own state
+machines.
+
+The state machine library used by the project will be robot3, which uses XState
+as an example, but chooses to be as lean as possible in its implementation.
+
+## Consequences
+
+1. Any view or component will come with a state machine to describe and manage
```

Comment:
Hmm I'm thinking it's a "document" that would have an associated state machine, and then present different views and components based on what's appropriate for its current state

Reply:
We discussed this in the channel just a few moments ago, but summarizing here for other reviewers: I actually *did* mean UI component, following the thoughts as expressed in https://xiaoyunyang.github.io/post/modeling-ui-state-using-a-finite-state-machine/.
I'm considering how to incorporate the nuances we discussed.

Reply:
So we agreed that there could be use-cases, but we need to be careful not to put too much (incomprehensible) boilerplate in place.